sql-plugin/src/main/scala/com/nvidia/spark/rapids/io/async/AsyncOutputStream.scala

```diff
@@ -0,0 +1,187 @@
+      case Some(t) => throw new IOException(t)
+      case None =>
+    }
+  }
+
+  @throws[IOException]
+  private def ensureOpen(): Unit = {
+    if (closed) {
+      throw new IOException("Stream closed")
+    }
+  }
+
+  override def write(b: Int): Unit = {
+    throwIfError()
+    ensureOpen()
+
+    val buffer = new Array[Byte](1)
+    buffer(0) = b.toByte
+    write(buffer)
+  }
+
+  /**
+   * Schedules a write of the given bytes to the underlying stream. The write is executed
+   * asynchronously on a background thread. The method returns immediately, and the write may not
+   * have completed when the method returns.
+   *
+   * If an error has occurred in the background thread and [[lastError]] has been set, this function
+   * will throw an IOException immediately.
+   *
+   * If an error has occurred in the background thread while executing a previous write after the
+   * current write has been scheduled, the current write will fail with the same error.
+   */
+  @throws[IOException]
+  override def write(b: Array[Byte], off: Int, len: Int): Unit = {
+    throwIfError()
+    ensureOpen()
+
+    metrics.numBytesScheduled += len
+    executor.submit(() => {
+      throwIfError()
+      ensureOpen()
+
+      try {
+        delegate.write(b, off, len)
+        metrics.numBytesWritten.addAndGet(len)
+      } catch {
+        case t: Throwable =>
+          // Update the error state
+          lastError.set(Some(t))
+      }
+    }, len)
+  }
+
+  /**
+   * Flushes all pending writes to the underlying stream. This method blocks until all pending
+   * writes have been completed. If an error has occurred in the background thread, this method
+   * will throw an IOException.
+   *
+   * If an error has occurred in the background thread and [[lastError]] has been set, this function
+   * will throw an IOException immediately.
+   *
+   * If an error has occurred in the background thread while executing a previous task after the
+   * current flush has been scheduled, the current flush will fail with the same error.
+   */
+  @throws[IOException]
+  override def flush(): Unit = {
+    throwIfError()
+    ensureOpen()
+
+    val f = executor.submit(() => {
+      throwIfError()
+      ensureOpen()
+
+      delegate.flush()
+    }, 0)
+
+    f.get()
+  }
+
+  /**
+   * Closes the underlying stream and releases any resources associated with it. All pending writes
+   * are flushed before closing the stream. This method blocks until all pending writes have been
+   * completed.
+   *
+   * If an error has occurred while flushing, this function will throw an IOException.
+   *
+   * If an error has occurred while executing a previous task before this function is called,
+   * this function will throw the same error. All resources and the underlying stream are still
+   * guaranteed to be closed.
+   */
+  @throws[IOException]
+  override def close(): Unit = {
+    if (!closed) {
+      Seq[AutoCloseable](
```

Comment:
I am not 100% sure how this works. It seems you are building a sequence of `AutoCloseable` but the lambdas are not `AutoCloseable`. `delegate` is, but that's it. What magic do you speak of?

Reply:
Ok discussed with @revans2 a bit it and he thought what is likely happening is yes that these are single method lambdas, and `AutoCloseable` wants an impl for a single method, so each lambda becomes:
```
Seq(
new AutoCloseable {
  override def close(): Unit = {
    flush()
  }
},
new AutoCloseable {
  override def close(): Unit = {
    executor.shutdownNow(10, TimeUnit.SECONDS)
  }
},
delegate,
new AutoCloseable {
  override def close(): Unit = {
    closed = true
  }
}).safeClose()
```

Note that `safeClose` will iterate through all of these, and call close in order. If _any_ throw, they get added as suppressed exceptions and thrown when the whole thing finishes. The effect will be that `AsyncOutputStream` will be closed, and we'll get an exception. I think that's all OK and it should take the task down (I don't know if the executor will come down.. but possibly it should?)

The main feedback is to add comments around here on how this works. You could also consider making an interface out of this, so it's clear what the intent is.

Reply:
I'm not sure if I understand your comments. So, I tried to explain the error propagation mechanism in this scaladoc [here](https://github.com/NVIDIA/spark-rapids/pull/11730/files#diff-de6876353f9d13e52ddacfb74d5229b259197690eb2bd20320b0ea069312adf8R159-R169). Do you think we need additional comments for this `close` function? Or do you think we need comments for `safeClose()`? Also what kind of interface are you suggesting?